factor/token/classes/factor.php

```diff
@@ -0,0 +1,251 @@
+        if ($verified === secret_manager::NONVALID) {
+            $this->set_state(\tool_mfa\plugininfo\factor::STATE_LOCKED);
+            return \tool_mfa\plugininfo\factor::STATE_LOCKED;
+        } else if ($verified === secret_manager::VALID) {
+            return \tool_mfa\plugininfo\factor::STATE_PASS;
+        }
+
+        // We should never get here. Factor cannot be revoked.
+        return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+    }
+
+    /**
+     * Token Implementation.
+     * We can't get_state like the parent here or it will recurse forever.
+     *
+     * @param string $state
+     * @return void
+     */
+    public function set_state($state) {
+        global $SESSION;
+        $property = 'factor_'.$this->name;
+        $SESSION->$property = $state;
+        return true;
+    }
+
+    /**
+     * Token implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function possible_states($user) {
+        return [
+            \tool_mfa\plugininfo\factor::STATE_PASS,
+            \tool_mfa\plugininfo\factor::STATE_NEUTRAL,
+            \tool_mfa\plugininfo\factor::STATE_LOCKED
+        ];
+    }
+
+    /**
+     * Token implementation.
+     * Inject a checkbox into every auth form if needed.
+     *
+     * {@inheritDoc}
+     */
+    public function global_definition_after_data($mform) {
+        global $SESSION;
+
+        // First thing, we need to decide on whether we should show the checkbox.
+        $noproperty = !property_exists($SESSION, 'tool_mfa_factor_token');
+        $nostate = $this->get_state() !== \tool_mfa\plugininfo\factor::STATE_PASS;
+
+        if ($noproperty && $nostate) {
+            $expiry = get_config('factor_token', 'expiry');
+            $expirystring = format_time($expiry);
+            $mform->addElement('advcheckbox', 'factor_token_trust', '', get_string('form:trust', 'factor_token', $expirystring));
+            $mform->setType('factor_token_trust', PARAM_BOOL);
+            $mform->setDefault('factor_token_trust', true);
+        }
+    }
+
+    /**
+     * Token implementation.
+     * Store information about the token status.
+     *
+     * {@inheritDoc}
+     */
+    public function global_submit($data) {
+        global $SESSION;
+
+        // Store any kind of response here, we shouldnt show again.
+        $trust = $data->factor_token_trust;
+        $SESSION->tool_mfa_factor_token = $trust;
+    }
+
+    /**
+     * Token implementation.
+     * Pass hook to set the cookie for use in subsequent auths.
+     *
+     * {@inheritDoc}
+     */
+    public function post_pass_state() {
+        global $SESSION, $USER;
+
+        if (!property_exists($SESSION, 'tool_mfa_factor_token')) {
+            return;
+        }
+        $settoken = $SESSION->tool_mfa_factor_token;
+        if (!$settoken) {
+            return;
+        }
+
+        $userid = $USER->id;
+        $cookie = 'MFA_TOKEN_' . $userid;
```

Comment:
```suggestion
        $cookie = 'MFA_TOKEN_' . $USER->id;
```

Should just use the variable directly since it's only used once in this function